### DISCUSSION TITLE
FakeFileWrapper: How handle `isinstance(ffw, io.BaseIO)`?

### DISCUSSION THREAD

**Discussion by buhtz** [jmcgeheeiv/pyfakefs#896] [Limitations] [answered]
Out there in the world packages and functions exist that checking the type of an object to find out if iti s a file-like object or a path/pathname.

When I do `pathlib.Path('foobar').open()` in a PyFakeFS environment the result is a `FakeFileWrapper`.
It behaves like a file-like object it didn't pass type tests like this.

`isinstance(file_object, io.BaseIO)`

Now the question is how to handle that?

**Comment by mrbean-bremen** [comment]
Good question - currently this is not possible. From the [documentation](http://jmcgeheeiv.github.io/pyfakefs/master/intro.html#limitations):
> pyfakefs does not retain the MRO for file objects, so you cannot rely on checks using isinstance for these objects (for example, to differentiate between binary and textual file objects).

This came up before (hence the note in the limitations), and so far I have no solution for this.
I did some instance faking for `pathlib.Path` fake classes though, and while it is not possible to do it the same way for `FakeFileWrapper`, there may be some other way to do this. I will at least have another go at this, though that may take some time.

**Comment by mrbean-bremen** [comment]
I'm moving this to discussions. I once already attempted to refactor the code accordingly, but got nowhere, so it is unlikely that this behavior will be changed any time soon.

**Comment by mrbean-bremen** [comment, marked as answer]
Note that I have recently changed the behavior, so that `FakeFileWrapper` now derives from `io.BufferedIOBase` or `io.TextIOBase` (it forwards all calls to the fake wrapper anyway). The way I had tried to refactor this before was too complicated - the current implementation basically just fakes the inheritance, so that `isinstance(file_object, io.BaseIO)` will now return `True`.
